File: Bloodborne-Blocks/tools/logical_contract_v2.py
"""Small authored schema-v2 adapter. No model elements participate in gameplay.

Java LogicalTransform and this module execute the same matrix contract/vectors.
Cells rotate around the master cell; points rotate around the explicit pivot.
Already oriented mesh/state data must never be rotated for a second time.
"""
from __future__ import annotations

import itertools
import json
import math
from pathlib import Path

BUDGETS = {"NONE": 0, "SIMPLE_BOX": 1, "TWO_BOX": 2, "THREE_BOX": 3, "TRUNK": 2,
           "POST": 1, "FENCE": 5, "WALL": 5, "DOOR": 2, "GATE": 2, "STAIRS": 3,
           "DECK": 9}
COLLISION_EPSILON = 1e-6
MATRICES = {"0": [[1,0,0],[0,1,0],[0,0,1]], "90": [[0,0,-1],[0,1,0],[1,0,0]],
            "180": [[-1,0,0],[0,1,0],[0,0,-1]], "270": [[0,0,1],[0,1,0],[-1,0,0]]}
POC_FAMILIES = {"o_dead_tree_planter", "o_cases_0", "o_wall_deco_1", "o_iron_gate", "o_iron_railing"}


def cell(value):
    if not isinstance(value, (list, tuple)) or len(value) != 3 or any(type(v) is not int or abs(v) > 64 for v in value):
        raise ValueError("expected bounded integer cell")
    return tuple(value)


def rotate_cell(value, rotation, transform):
    matrix = transform["rotations"][str(rotation)]
    return tuple(sum(row[i] * value[i] for i in range(3)) for row in matrix)


def master_origin(placement_cell, anchor_cell, rotation, transform):
    offset = rotate_cell(anchor_cell, rotation, transform)
    return tuple(placement_cell[i] - offset[i] for i in range(3))


def rotate_box(box, rotation, transform, pivot=(.5, 0, .5)):
    points = [tuple(rotate_cell(tuple(p[i]-pivot[i] for i in range(3)), rotation, transform)[j]+pivot[j]
                    for j in range(3)) for p in itertools.product(*zip(box[:3], box[3:]))]
    return [min(p[i] for p in points) for i in range(3)] + [max(p[i] for p in points) for i in range(3)]


def box_cells(box):
    return set(itertools.product(*(range(math.floor(box[i]+COLLISION_EPSILON),
                                         math.ceil(box[i+3]-COLLISION_EPSILON)) for i in range(3))))


def check_box(box):
    if not isinstance(box, list) or len(box) != 6 or any(type(n) not in (int, float) or not math.isfinite(n) or abs(n)>64 for n in box):
        raise ValueError("invalid bounded box")
    if any(box[i] >= box[i+3] for i in range(3)):
        raise ValueError("empty gameplay box")


def load_transform(path):
    value = json.loads(Path(path).read_text(encoding="utf-8"))
    if value.get("schemaVersion") != 2 or value.get("rotations") != MATRICES or len(value.get("vectors", [])) < 4:
        raise ValueError("invalid transform contract")
    for vector in value["vectors"]:
        if master_origin(vector["placement_cell"], cell(vector["anchor_cell"]), vector["rotation"], value) != tuple(vector["expected_master"]):
            raise ValueError("transform vector mismatch")
    return value


def load_contracts(resources):
    resources = Path(resources)
    data = json.loads((resources / "contracts-v2.json").read_text(encoding="utf-8"))
    if data.get("schemaVersion") != 2 or data.get("transform_contract") != "transform-v2.json":
        raise ValueError("invalid logical contract schema")
    transform = load_transform(resources / data["transform_contract"])
    production_path = resources / 'production-palette.json'
    if production_path.exists():
        production_ids = {o['id'] for o in json.loads(production_path.read_text(encoding='utf8'))['objects']}
        if production_ids != {f['id'] for f in data['families']}:
            raise ValueError('production manifest/contract mismatch')
    elif not POC_FAMILIES <= {f["id"] for f in data["families"]}:
        raise ValueError("schema-v2 must preserve the approved POC families")
    definitions = {d["id"]: d for d in json.loads((resources / "definitions.json").read_text(encoding="utf-8"))["blocks"]}
    physical_path = resources / 'physical-footprints.json'
    physical = json.loads(physical_path.read_text(encoding='utf-8')) if physical_path.exists() else None
    if physical is not None and (physical.get('schemaVersion') != 1 or
            set(physical.get('families', {})) != {f['id'] for f in data['families']}):
        raise ValueError('physical footprint family coverage')
    seen = set()
    split_patterns = {}
    for family in data["families"]:
        ident = family["id"]
        if ident in seen or ident not in definitions or not definitions[ident].get("logical"):
            raise ValueError("duplicate/unknown family")
        seen.add(ident)
        if not isinstance(family.get("migration_disabled", False), bool):
            raise ValueError("migration_disabled must be boolean")
        if family.get('authority') not in (None,'user') or bool(family.get('review_id')) != bool(family.get('authority')):
            raise ValueError('invalid authoritative review metadata')
        cell(family["canonical_anchor"]["cell"])
        if family["canonical_anchor"]["pivot"] != [.5,0,.5] or family["rotations"] != [0,90,180,270] or family["mirror_policy"] != "ROTATE_ONLY":
            raise ValueError("unsupported transform policy")
        if family["placement_policy"] not in ("FLOOR", "WALL_ADJACENT"):
            raise ValueError("unsupported placement policy")
        policy = family["collision_policy"]
        budget = BUDGETS[policy]
        if policy in ("TRUNK", "STAIRS", "TWO_BOX", "THREE_BOX", "DECK") and not family.get("collision_justification"):
            raise ValueError("multiple primitive policy needs justification")
        if set(family["states"]) != set(definitions[ident]["states"]):
            raise ValueError("contract must describe all existing family states")
        if physical is not None and set(physical['families'][ident]) != set(family['states']):
            raise ValueError('physical footprint state coverage')
        for key, state in family["states"].items():
            mask = physical['families'][ident][key] if physical is not None else {
                'cells': state['interaction_footprint']['cells'], 'boxes': state['collision_footprint']['boxes']}
            physical_cells = [cell(c) for c in mask['cells']]
            if len(physical_cells) != len(set(physical_cells)) or (0,0,0) not in physical_cells or len(physical_cells)>512:
                raise ValueError('invalid physical footprint')
            if len(mask['boxes']) > budget:
                raise ValueError('physical collision budget exceeded')
            for box in mask['boxes']:
                check_box(box)
                if not box_cells(box) <= set(physical_cells):
                    raise ValueError('COLLISION_OUTSIDE_PHYSICAL_FOOTPRINT')
            if family['placement_policy']=='FLOOR' and any(c[1]<0 for c in physical_cells):
                raise ValueError('physical helper below floor anchor')
            state['physical_footprint'] = mask
            if state["rotation"] not in family["rotations"] or state["render_mesh"]["id"] != definitions[ident]["models"][key]:
                raise ValueError("state/mesh mismatch")
            render = state["render_mesh"]
            if len(render["bounds"]) != 6 or len(render["offset"]) != 3 or any(not math.isfinite(n) or abs(n)>64 for n in render["bounds"]+render["offset"]):
                raise ValueError("invalid render metadata")
            selection, collision = state["selection_footprint"]["boxes"], state["collision_footprint"]["boxes"]
            if len(selection) != 1 or len(collision) > budget:
                raise ValueError("primitive budget exceeded")
            for box in selection+collision:
                check_box(box)
            cells = [cell(c) for c in state["interaction_footprint"]["cells"]]
            if len(cells) != len(set(cells)) or (0,0,0) not in cells or len(cells)>512:
                raise ValueError("invalid explicit interaction footprint")
            if family['placement_policy']=='FLOOR':
                if any(c[1]<0 for c in cells):
                    raise ValueError('GROUND_OBJECT_HAS_HELPER_BELOW_ANCHOR: '+ident)
                if render['bounds'][1]+render['offset'][1] < -1e-6:
                    raise ValueError('RENDER_BELOW_SUPPORT_PLANE: '+ident)
            if any(not box_cells(b) <= set(cells) for b in collision):
                raise ValueError("COLLISION_OUTSIDE_OWNED_CELLS: "+ident+"["+key+"]")
            for pattern in state["migration_source_pattern"]:
                components = pattern["components"]
                positions = [cell(c["offset"]) for c in components]
                if not components or len(positions) != len(set(positions)):
                    raise ValueError("empty or overlapping source pattern")
                for component in components:
                    if not component["id"].startswith("minecraft:") or not isinstance(component["properties"], dict):
                        raise ValueError("direct pattern must describe raw minecraft states")
                for guard in pattern.get('variant_guards',[]):
                    if cell(guard['offset']) not in positions or type(guard.get('multipart')) is not bool:
                        raise ValueError('RNG guard must reference a matched source cell')
                    weights,indices=guard['weights'],guard['indices']
                    if not weights or any(type(w) is not int or w<=0 for w in weights) or sum(weights)>2147483647:
                        raise ValueError('invalid RNG weights')
                    if not indices or len(set(indices))!=len(indices) or any(type(i) is not int or not 0<=i<len(weights) for i in indices):
                        raise ValueError('invalid RNG choices')
                transaction = pattern.get("split_transaction")
                if transaction is not None:
                    if (not isinstance(transaction, dict) or set(transaction) != {"id", "outputs"} or
                            not isinstance(transaction["id"], str) or not transaction["id"] or len(transaction["id"]) > 128 or
                            not isinstance(transaction["outputs"], list) or len(transaction["outputs"]) < 2 or
                            any(not isinstance(output, dict) or not {"family", "root_offset"} <= set(output) or set(output)-{"family", "root_offset", "properties"} or
                                not isinstance(output["family"], str) or not output["family"] for output in transaction["outputs"])):
                        raise ValueError("invalid split transaction")
                    outputs = tuple((output["family"], cell(output["root_offset"]), tuple(sorted(output.get('properties',{}).items()))) for output in transaction["outputs"])
                    for output in transaction['outputs']:
                        target_def=definitions.get(output['family'])
                        override=output.get('properties',{})
                        if not target_def or any(k not in target_def['properties'] or v not in target_def['properties'][k] for k,v in override.items()):
                            raise ValueError('invalid split output state override')
                    # One rendered family may legitimately be placed more than once
                    # (for example two identical statues at different roots).  The
                    # declaration identity is therefore family + root, while the
                    # shared raw pattern still has exactly one member per family.
                    if len(set(outputs)) != len(outputs) or ident not in {output[0] for output in outputs}:
                        raise ValueError("split transaction outputs must have unique family/root pairs and include family")
                    signature = (tuple(sorted((cell(component["offset"]), component["id"],
                                               tuple(sorted(component["properties"].items()))) for component in components)),
                                 json.dumps(pattern.get("variant_guards", []), sort_keys=True, separators=(",", ":")))
                    split_patterns.setdefault((transaction["id"], signature), []).append((ident, outputs))
    known = {family["id"] for family in data["families"]}
    for (_, _), members in split_patterns.items():
        declared = members[0][1]
        declared_ids = {output[0] for output in declared}
        member_ids = [ident for ident, _ in members]
        if (any(outputs != declared for _, outputs in members) or
                len(member_ids) != len(declared_ids) or len(set(member_ids)) != len(member_ids) or
                declared_ids != set(member_ids) or declared_ids - known):
            raise ValueError("split transaction must declare exactly one shared source pattern per output")
    return data, transform


def direct_rules(resources, *, poc_only=False):
    """Compile only explicit vanilla patterns into the existing safe transaction engine."""
    from convert_logical_world import Expected, Output, Rule, add, load_defaults, make_state
    data, transform = load_contracts(resources)
    defaults = load_defaults(resources)
    rules = []
    split = {}
    for family in data["families"]:
        if poc_only and family['id'] not in POC_FAMILIES:
            continue
        if family.get("migration_disabled", False) and not poc_only:
            continue
        # QA2 explicitly retires the incomplete tree construction patterns.
        # Keep frozen POC mode for historical regression fixtures only. In the
        # current converter an incomplete tree must never fall back to a wing
        # or a trunk-only object when its full sixteen-cell pattern is absent.
        if not poc_only and any(f['id'] == 'o_c001' for f in data['families']) and family['id'] in {
                'o_c001_a', 'o_c001_b', 'o_c009_a', 'o_c009_b', 'o_dead_tree_planter'}:
            continue
        for key, state in family["states"].items():
            properties = dict(part.split("=", 1) for part in key.split(",") if part)
            target = make_state({"id": family["id"], "properties": properties}, defaults)
            for pattern in state["migration_source_pattern"]:
                # Source offsets are already rotated and relative to the canonical master.
                # The seed need not be at the master; no min-cell heuristics are used.
                pieces = [Expected(cell(c["offset"]), make_state(c, {})) for c in pattern["components"]]
                first = pieces[0]
                # candidates subtracts first.offset to obtain master. Evaluate the same
                # explicit-anchor transform used by item placement (zero relative shift).
                anchor = rotate_cell(family["canonical_anchor"]["cell"], state["rotation"], transform)
                shift = master_origin(anchor, family["canonical_anchor"]["cell"], state["rotation"], transform)
                shape = frozenset(cell(c) for c in state["physical_footprint"]["cells"])
                transaction = pattern.get("split_transaction")
                if transaction is None:
                    rules.append(Rule(len(rules), first, target, shift, tuple(pieces[1:]), None, shape,
                                      supersedes_targets=(frozenset('bloodborne_blocks:'+f['id'] for f in data['families'] if f['id']!=family['id']) if family.get('authority')=='user' else frozenset('bloodborne_blocks:'+i for i in family.get('supersedes_targets',[]))),
                                      variant_guards=tuple(pattern.get('variant_guards', []))))
                    continue
                signature = (tuple(sorted((piece.offset, piece.state) for piece in pieces)),
                             json.dumps(pattern.get("variant_guards", []), sort_keys=True, separators=(",", ":")))
                split.setdefault((transaction["id"], signature), []).append((family["id"], target, shift, shape, pieces, tuple(pattern.get("variant_guards", [])),
                                                                               tuple((output["family"], cell(output["root_offset"]), tuple(sorted(output.get('properties',{}).items()))) for output in transaction["outputs"])))
    for (transaction_id, _), members in split.items():
        members.sort(key=lambda member: member[0])
        first_family, target, shift, shape, pieces, guards, declarations = members[0]
        by_family = {member[0]: member for member in members}
        declared_families = {family for family, _, _ in declarations}
        member_families = [member[0] for member in members]
        if len(by_family) != len(member_families):
            raise ValueError("split transaction must declare exactly one shared source pattern per output")
        if declared_families != set(by_family):
            # A disabled historical composite makes its declared split incomplete.
            # Do not silently degrade to the remaining outputs.
            continue
        outputs_list=[]
        family_map={f['id']:f for f in data['families']}
        for family,root_offset,override in declarations:
            base=by_family[family]
            props={**dict(base[1][1]),**dict(override)}
            target_state=make_state({'id':family,'properties':props},defaults)
            state_key=','.join(f'{k}={v}' for k,v in sorted(props.items()))
            target_shape=frozenset(cell(c) for c in family_map[family]['states'][state_key]['physical_footprint']['cells'])
            outputs_list.append(Output(target_state,add(base[2],root_offset),target_shape))
        outputs=tuple(outputs_list)
        occupied = set()
        for output in outputs:
            cells = {tuple(output.root_offset[i] + offset[i] for i in range(3)) for offset in output.shape}
            if occupied & cells:
                raise ValueError("split transaction output footprints overlap")
            occupied.update(cells)
        rules.append(Rule(len(rules), pieces[0], target, shift, tuple(pieces[1:]), None, shape,
                          variant_guards=guards, transaction_id=transaction_id, outputs=outputs))
    if not poc_only:
        # A newer user-reviewed exact raw pattern is authoritative over the old
        # POC matcher. Existing POC blocks/assets remain untouched. In particular
        # C046 is proven against 65 original polygons, not the POC's 71 polygons.
        from dataclasses import replace
        def raw_signature(rule):
            return (tuple(sorted((p.offset,p.state) for p in (rule.source,)+rule.members)),
                    json.dumps(rule.variant_guards,sort_keys=True))
        reviewed_ids={f['id'] for f in data['families'] if f.get('authority')=='user'}
        reviewed={raw_signature(r) for r in rules if r.target[0].split(':',1)[1] in reviewed_ids}
        rules=[r for r in rules if r.target[0].split(':',1)[1] not in POC_FAMILIES-reviewed_ids or raw_signature(r) not in reviewed]
        rules=[replace(rule,number=i) for i,rule in enumerate(rules)]
    return rules, defaults

File: Bloodborne-Blocks/src/main/java/dev/dreamwalker/bloodborneblocks/LogicalContractV2.java
package dev.dreamwalker.bloodborneblocks;

import com.google.gson.Gson;
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.*;

/** Strict loader for the small authored logical-object contract (schema v2). */
final class LogicalContractV2 {
 private static final Gson GSON=new Gson();
 private static final Set<String> REQUIRED_POC_FAMILIES=Set.of("o_dead_tree_planter","o_cases_0","o_wall_deco_1","o_iron_gate","o_iron_railing");
 private static Map<String,DebugMetadata> DEBUG_METADATA=Map.of();
 static final class Data {int schemaVersion;String transform_contract;List<Family> families;}
 static final class Family {String id,placement_policy,mirror_policy,collision_policy,selection_policy,collision_justification,review_id,authority;Anchor canonical_anchor;List<Integer> rotations;Map<String,State> states;List<Pattern> migration_source_pattern,review_source_patterns;}
 static final class Anchor {int[] cell;double[] pivot;}
 static final class State {int rotation;Render render_mesh;Footprint selection_footprint,collision_footprint,interaction_footprint;List<Pattern> migration_source_pattern;}
 static final class Render {String id;double[] bounds,offset;}
 static final class Footprint {List<double[]> boxes,cells;}
 static final class Pattern {String exact_source_signature;List<Component> components;}
 static final class Component {String id;Map<String,String> properties;int[] offset;}
 record DebugMetadata(String reviewId,String canonicalAnchor,String collisionPolicy,String selectionPolicy,String sourceFamily,String sourcePatternSummary) {}
 private LogicalContractV2() {}

 static Map<String,GeometryRuntime.GeometryBlock> load(BloodborneBlocks.Data definitions){
  LogicalTransform.loadAndValidate();Data data=read();if(data.schemaVersion!=2||!"transform-v2.json".equals(data.transform_contract)||data.families==null)throw fail("schema");
  Map<String,BloodborneBlocks.Definition> known=new HashMap<>();for(BloodborneBlocks.Definition d:definitions.blocks)known.put(d.id,d);Set<String> meshes=ModularMeshData.loadLogicalAndValidate().keySet();
  Map<String,GeometryRuntime.GeometryBlock> result=new HashMap<>();Map<String,DebugMetadata> debug=new HashMap<>();Set<String> ids=new HashSet<>();
  for(Family family:data.families){
   if(family==null||family.id==null||!ids.add(family.id)||family.states==null||family.canonical_anchor==null||family.rotations==null)throw fail("family");
   BloodborneBlocks.Definition definition=known.get(family.id);if(definition==null||!definition.logical)throw fail("unknown family "+family.id);
   checkAnchor(family.canonical_anchor);if(!Set.of("FLOOR","WALL_ADJACENT").contains(family.placement_policy)||!"ROTATE_ONLY".equals(family.mirror_policy)||!family.rotations.equals(List.of(0,90,180,270)))throw fail("policy "+family.id);
   if(family.id.startsWith("o_c")&&!Arrays.equals(family.canonical_anchor.cell,new int[]{0,0,0}))throw fail("batch anchor "+family.id);
   if((family.review_id==null)!=(family.authority==null)||(family.review_id!=null&&(family.review_id.isBlank()||!"user".equals(family.authority))))throw fail("review metadata "+family.id);
   if(!family.states.keySet().equals(definition.states.keySet())||definition.models==null)throw fail("states "+family.id);
   checkOrientations(family);
   GeometryRuntime.GeometryBlock block=new GeometryRuntime.GeometryBlock();block.states=new HashMap<>();
   for(var entry:family.states.entrySet()){
    String key=entry.getKey();State state=entry.getValue();if(state==null||!family.rotations.contains(state.rotation)||state.render_mesh==null||!Objects.equals(definition.models.get(key),state.render_mesh.id)||!meshes.contains(state.render_mesh.id))throw fail("mesh "+family.id+"["+key+"]");
    checkRender(state.render_mesh);checkBoxes(state.selection_footprint,"selection",1);checkBoxes(state.collision_footprint,"collision",budget(family));checkCells(state.interaction_footprint);checkPattern(state.migration_source_pattern);
    if("FLOOR".equals(family.placement_policy)){
     if(state.render_mesh.bounds[1]+state.render_mesh.offset[1]<-1e-6)throw fail("RENDER_BELOW_SUPPORT_PLANE "+family.id+"["+key+"]");
     for(double[] cell:state.interaction_footprint.cells)if(cell[1]<0)throw fail("GROUND_OBJECT_HAS_HELPER_BELOW_ANCHOR "+family.id+"["+key+"]");
    }
    for(double[] box:state.collision_footprint.boxes)if(!covered(box,state.interaction_footprint.cells))throw fail("collision cell coverage "+family.id+"["+key+"]");
    block.states.put(key,geometry(family,state));
    debug.put(family.id+"\u0000"+key,new DebugMetadata(family.review_id,anchorText(family.canonical_anchor),family.collision_policy,family.selection_policy==null?"AUTHORED_OUTLINE":family.selection_policy,sourceFamily(family),sourcePatternSummary(family,state)));
   }
   result.put(family.id,block);
  }
  if(!result.keySet().containsAll(REQUIRED_POC_FAMILIES))throw fail("missing approved POC family");
  DEBUG_METADATA=Collections.unmodifiableMap(debug);
  return result;
 }
 static DebugMetadata debugMetadata(String familyId,String stateKey){return DEBUG_METADATA.get(familyId+"\u0000"+stateKey);}
 private static String anchorText(Anchor anchor){return anchor.cell[0]+" "+anchor.cell[1]+" "+anchor.cell[2]+" (local cell; pivot 0.5 0 0.5)";}
 private static String sourceFamily(Family family){return "catalog review "+(family.review_id==null?"noCatalogID":family.review_id)+"; placed provenance not stored";}
 private static String sourcePatternSummary(Family family,State state){LinkedHashSet<String> signatures=new LinkedHashSet<>();if(state.migration_source_pattern!=null)for(Pattern pattern:state.migration_source_pattern)if(pattern!=null&&pattern.exact_source_signature!=null&&!pattern.exact_source_signature.isBlank())signatures.add(pattern.exact_source_signature.substring(0,Math.min(12,pattern.exact_source_signature.length())));String listed=signatures.stream().limit(3).reduce((left,right)->left+","+right).orElse("none");return "catalog only; state signatures="+listed+", total="+signatures.size()+", family patterns="+count(family.review_source_patterns)+"; placed provenance not stored";}
 private static int count(List<Pattern> patterns){return patterns==null?0:patterns.size();}
 static Set<String> declaredFamilyIds(){Data data=read();if(data.schemaVersion!=2||data.families==null)throw fail("schema");Set<String> ids=new LinkedHashSet<>();for(Family family:data.families)if(family==null||family.id==null||!ids.add(family.id))throw fail("family");return Collections.unmodifiableSet(ids);}
 private static void checkOrientations(Family family){
  Map<String,Map<String,State>> oriented=new HashMap<>();
  Map<String,Map<String,State>> connections=new HashMap<>();
  for(var entry:family.states.entrySet()){
   Map<String,String> properties=parseStateKey(entry.getKey());String facing=properties.remove("facing");
   if(facing==null&&isConnectionMask(properties)){
    String group="";Map<String,String> remainder=new TreeMap<>(properties);for(String direction:List.of("north","east","south","west"))remainder.remove(direction);
    if(!remainder.isEmpty())group=remainder.entrySet().stream().map(e->e.getKey()+"="+e.getValue()).reduce((a,b)->a+","+b).orElse("");
    if(connections.computeIfAbsent(group,ignored->new HashMap<>()).put(connectionMask(properties),entry.getValue())!=null)throw fail("duplicate connection mask "+family.id);continue;
   }
   String group=properties.entrySet().stream().sorted(Map.Entry.comparingByKey()).map(e->e.getKey()+"="+e.getValue()).reduce((a,b)->a+","+b).orElse("");
   if(facing==null){if(entry.getValue().rotation!=0)throw fail("symmetric rotation "+family.id+"["+entry.getKey()+"]");continue;}
   if(!Set.of("north","east","south","west").contains(facing))throw fail("facing "+family.id+"["+entry.getKey()+"]");
   if(oriented.computeIfAbsent(group,ignored->new HashMap<>()).put(facing,entry.getValue())!=null)throw fail("duplicate facing "+family.id+"["+entry.getKey()+"]");
  }
  for(var entry:oriented.entrySet()){
   if(!entry.getValue().keySet().equals(Set.of("north","east","south","west")))throw fail("incomplete rotations "+family.id);
   for(var expected:Map.of("north",0,"east",90,"south",180,"west",270).entrySet())if(entry.getValue().get(expected.getKey()).rotation!=expected.getValue())throw fail("rotation "+family.id+" "+expected.getKey());
  }
  for(var entry:connections.entrySet()){
   if(entry.getValue().size()!=16)throw fail("incomplete connection masks "+family.id);
   for(int mask=0;mask<16;mask++){
    State state=entry.getValue().get(Integer.toString(mask));if(state==null||state.rotation!=0)throw fail("connection rotation "+family.id);
    int rotated=((mask&1)<<1)|((mask&2)<<1)|((mask&4)<<1)|((mask&8)>>>3);if(!entry.getValue().containsKey(Integer.toString(rotated)))throw fail("connection rotation closure "+family.id);
   }
  }
 }
 private static boolean isConnectionMask(Map<String,String> properties){return properties.keySet().containsAll(Set.of("north","east","south","west"))&&properties.entrySet().stream().filter(e->Set.of("north","east","south","west").contains(e.getKey())).allMatch(e->Set.of("false","true").contains(e.getValue()));}
 private static String connectionMask(Map<String,String> properties){int mask=0;for(int index=0;index<4;index++)if("true".equals(properties.get(List.of("north","east","south","west").get(index))))mask|=1<<index;return Integer.toString(mask);}
 private static Map<String,String> parseStateKey(String key){
  if(key==null)throw fail("null state key");Map<String,String> result=new HashMap<>();if(key.isEmpty())return result;
  for(String entry:key.split(",")){String[] pair=entry.split("=",-1);if(pair.length!=2||pair[0].isBlank()||pair[1].isBlank()||result.put(pair[0],pair[1])!=null)throw fail("state key "+key);}return result;
 }
 private static Data read(){try(InputStream in=LogicalContractV2.class.getResourceAsStream("/bloodborne_blocks/logical/contracts-v2.json")){if(in==null)throw fail("missing contracts-v2.json");return GSON.fromJson(new InputStreamReader(in,StandardCharsets.UTF_8),Data.class);}catch(IOException|RuntimeException e){throw e instanceof IllegalStateException?(IllegalStateException)e:new IllegalStateException("Cannot load contracts-v2",e);}}
 private static GeometryRuntime.GeometryState geometry(Family family,State state){
  GeometryRuntime.GeometryState geometry=new GeometryRuntime.GeometryState();geometry.anchor=family.canonical_anchor.cell.clone();geometry.rotation=state.rotation;geometry.render_offset=state.render_mesh.offset.clone();geometry.cells=new LinkedHashMap<>();
  geometry.placementPolicy=family.placement_policy;geometry.mirrorPolicy=family.mirror_policy;
  for(double[] cell:state.interaction_footprint.cells){int x=(int)cell[0],y=(int)cell[1],z=(int)cell[2];GeometryRuntime.GeometryCell part=new GeometryRuntime.GeometryCell();part.collision=clip(state.collision_footprint.boxes,x,y,z);part.outline=clip(state.selection_footprint.boxes,x,y,z);geometry.cells.put(x+","+y+","+z,part);}
  geometry.globalOutline=state.selection_footprint.boxes.get(0).clone();return geometry;
 }
 private static List<double[]> clip(List<double[]> boxes,int x,int y,int z){List<double[]> out=new ArrayList<>();for(double[] box:boxes){double[] clipped={Math.max(box[0],x)-x,Math.max(box[1],y)-y,Math.max(box[2],z)-z,Math.min(box[3],x+1)-x,Math.min(box[4],y+1)-y,Math.min(box[5],z+1)-z};if(clipped[0]<clipped[3]&&clipped[1]<clipped[4]&&clipped[2]<clipped[5])out.add(clipped);}return out;}
 private static boolean covered(double[] box,List<double[]> cells){for(int x=(int)Math.floor(box[0]);x<(int)Math.ceil(box[3]);x++)for(int y=(int)Math.floor(box[1]);y<(int)Math.ceil(box[4]);y++)for(int z=(int)Math.floor(box[2]);z<(int)Math.ceil(box[5]);z++){boolean present=false;for(double[] cell:cells)if((int)cell[0]==x&&(int)cell[1]==y&&(int)cell[2]==z)present=true;if(!present)return false;}return true;}
 private static int budget(Family f){int n=switch(f.collision_policy){case "NONE"->0;case "SIMPLE_BOX"->1;case "TWO_BOX"->2;case "THREE_BOX"->3;case "TRUNK"->2;case "POST"->1;case "FENCE","WALL"->5;case "DOOR","GATE"->2;case "STAIRS"->3;default->-1;};if(n<0||(("TRUNK".equals(f.collision_policy)||"STAIRS".equals(f.collision_policy)||"TWO_BOX".equals(f.collision_policy)||"THREE_BOX".equals(f.collision_policy))&&(f.collision_justification==null||f.collision_justification.isBlank())))throw fail("collision policy "+f.id);return n;}
 private static void checkAnchor(Anchor a){LogicalTransform.requireCell(a.cell,"anchor");for(int n:a.cell)if(Math.abs(n)>64)throw fail("anchor cell");if(a.pivot==null||a.pivot.length!=3||a.pivot[0]!=.5||a.pivot[1]!=0||a.pivot[2]!=.5)throw fail("anchor pivot");}
 private static void checkRender(Render r){if(r.id==null||r.bounds==null||r.bounds.length!=6||r.offset==null||r.offset.length!=3)throw fail("render mesh");for(double n:r.bounds)bounded(n,"render bounds");for(double n:r.offset)bounded(n,"render offset");}
 private static void checkBoxes(Footprint f,String name,int max){if(f==null||f.boxes==null||f.boxes.size()>max||("selection".equals(name)&&f.boxes.size()!=1))throw fail(name+" budget");for(double[] b:f.boxes){if(b==null||b.length!=6)throw fail(name+" box");for(double n:b)bounded(n,name);if(b[0]>=b[3]||b[1]>=b[4]||b[2]>=b[5])throw fail(name+" empty");}}
 private static void checkCells(Footprint f){if(f==null||f.cells==null||f.cells.isEmpty())throw fail("interaction cells");Set<String> seen=new HashSet<>();boolean root=false;for(double[] c:f.cells){if(c==null||c.length!=3)throw fail("interaction cell");for(double n:c)if(!Double.isFinite(n)||n!=Math.rint(n)||Math.abs(n)>64)throw fail("interaction cell");if(c[0]==0&&c[1]==0&&c[2]==0)root=true;if(!seen.add((int)c[0]+","+(int)c[1]+","+(int)c[2]))throw fail("duplicate interaction cell");}if(!root)throw fail("interaction cells omit root");}
 private static void checkPattern(List<Pattern> patterns){if(patterns==null)return;for(Pattern p:patterns){if(p==null||p.components==null)throw fail("migration pattern");for(Component c:p.components){if(c==null||c.id==null||c.offset==null)throw fail("migration component");LogicalTransform.requireCell(c.offset,"migration offset");}}}
 private static void bounded(double n,String what){if(!Double.isFinite(n)||Math.abs(n)>64)throw fail(what);}
 private static IllegalStateException fail(String message){return new IllegalStateException("Invalid logical contract v2: "+message);}
}
